haskell-api.md

```diff
@@ -0,0 +1,217 @@
+# Haskell API Conventions
+
+## Stateless
+
+Prefer stateless endpoints. Sessions are used to authorize data access, but not
+choose what data to return.
+
+- Good:
+  - `/3/students?schools.id=x,y,z`
+  - `ApiClient.fetch("/3/students", { "schools.id": me.schoolIds })`
+- Bad:
+  - `/3/school-admins/me/students`
+  - `ApiClient.fetch("/3/school-admins/me/students")`
+
+## Parameters
+
+We want to :heart: query parameters as a way to keep the number of distinct
+endpoints lower by supporting more complex filtering combinations within each
+endpoints.
+
+- Good: `/3/students?schools.id=x,y,z&teachers.id=a,b,c`
+- Bad:
+  - `/3/schools/:id/students`
+  - `/3/teachers/:id/students`
+  - `/3/schools/:id/teachers/:id/students`
+  - etc, forever
+
+### Naming
+
+- If filtering on an attribute of the main resource: `{attribute}={value}`
+- If filtering on an attribute of a related resource:
+  `{resource}.{attribute}={value}`
```

Comment:
It seems like we consider foreign keys to be considered an attribute on a related resource. i.e.

```sql
CREATE TABLE teacher
  ( schoolId int REFERENCES schools)
```

Bad
```
GET /3/teachers?schoolId=123
```

Good
```
GET /3/teachers?schools.id=123
```